Feature(authentication): add Linux support using zbus and polkit

tyreseluo opened the pull request:
## PR content

Add Linux support using zbus and polkit for authentication crate and  clear the database errors under Linux.

The basic Linux authentication pop-up has been implemented, but some details still need to be confirmed.

<img width="429" height="378" alt="image" src="https://github.com/user-attachments/assets/931f2358-9db8-4249-a03d-714eefa2f98a" />

Still in development, with an update coming this week.

## PR Changes
- Replace polkit and gio dependencies with zbus and zbus_polkit
- Implement D-Bus-based authentication for Linux using polkit
- Add complete Linux implementation module with proper error handling
- Update example to handle Linux async authentication pattern
- Add comprehensive Linux usage documentation to README
- Enable Linux target in sys.rs module

tyreseluo committed:
feat(authentication): add Linux support using zbus and polkit

- Replace polkit and gio dependencies with zbus and zbus_polkit
- Implement D-Bus-based authentication for Linux using polkit
- Add complete Linux implementation module with proper error handling
- Add Tyrese Luo as co-author to authentication crate
- Update example to handle Linux async authentication pattern
- Add comprehensive Linux usage documentation to README
- Enable Linux target in sys.rs module

tyreseluo committed:
Authentication Linux crate, polkit action ID requirement and caching tweaks

  - note Linux action ID requirement in example and expose PolicyBuilder::action_id on Linux
  - cache system bus and polkit authority, name auth thread, refine authorization result handling
  - make authentication Error clonable for reuse in cached results

tyreseluo committed:
Add linux authentication screenshot

tyreseluo committed:
Update andclean up

kevinaboos commented on crates/authentication/README.md:
```diff
@@ -30,14 +30,80 @@ To use this crate on Android, you must add the following to your app's `AndroidM
+Log back into your desktop session (or restart polkitd), then run your example.
+
+
+Plicy file example:
```
```suggestion
Policy file example:
```

kevinaboos commented on crates/authentication/README.md:
```diff
@@ -30,14 +30,80 @@ To use this crate on Android, you must add the following to your app's `AndroidM
+#### 
+
+> [!TIP]
+> Automatic addition during runtime is not recommended and should generally be avoided.
```
what does "automatic addition" mean? Addition of what? please clarify.

kevinaboos commented on crates/authentication/README.md:
```diff
@@ -30,14 +30,80 @@ To use this crate on Android, you must add the following to your app's `AndroidM
     AndroidText, BiometricStrength, Context, Policy, PolicyBuilder, Text, WindowsText,
 };
 
+// Linux ignores policy options like biometrics/password (kept for parity).
```
```suggestion
// Linux ignores policy options like biometrics/password,
// but it's best to keep them in for proper operation on other platforms.
```

kevinaboos commented on crates/authentication/examples/simple_authentication.rs:
```diff
@@ -22,19 +25,47 @@ const TEXT: Text = Text {
+        }
+
+        return;
+    }
```
in general, the entire purpose of this crate is to allow the application to be cross-platform easily without needing to use cfg blocks. Is the linux way of doing things not compatible with the other platforms? If not, we need to redesign it such that a single platform-independent code block can work on all platforms. This is particularly important for the example code, which is what developers will look at as a reference when trying to understand our crate.

kevinaboos commented on crates/authentication/src/lib.rs:
```diff
@@ -199,6 +199,16 @@ impl PolicyBuilder {
         }
     }
 
+    #[inline]
+    #[must_use]
+    #[cfg(target_os = "linux")]
+    pub const fn action_id(self, id: &'static str) -> Self {
+        Self {
+            inner: self.inner.action_id(id),
+        }
+    }
```
again, the whole point of this crate is to provide a single top-level interface that is platform-agnostic.

That means that, in general, we don't want any platform-specific code or cfg blocks in the public API of this crate. 

If the concept of an "action ID" is mandatory in the top-level crate API, then one way to provide that is to just remove the cfg attribute and allow it to be called on all platforms, but just make it do nothing on platforms that it's irrelevant to. This is what we do for all other similar features; in fact, you can see an example of this right below this code, in the `biometrics()` fn, which only has an effect on Android.

When designing the API, think carefully about how you might _use_ it from a single cross-platform app like Robrix. That is, if I wanted to use this in Robrix with no cfg-specific code, how would I do so? Ensure that the crate API supports that key usage pattern.

kevinaboos commented on crates/authentication/src/sys/linux/mod.rs:
```diff
@@ -0,0 +1,169 @@
+
+    pub(crate) fn authenticate<F>(
+        &self,
+        _message: Text, // _message is unused on Linux, see reasons below:  auth.check_authorization();
```
this comment should be a doc comment on the top-level `authenticate` function such that a user of this crate can actually see it. It's quite important for the user of this crate to know that, such that they can correctly create a policy file that contains the proper authentication message string.

kevinaboos commented on crates/authentication/src/sys/linux/mod.rs:
```diff
@@ -0,0 +1,169 @@
+    {
+        let action_id = policy.action_id;
+        // Here, we perform a simple and direct thread creation because the current calls are infrequent.
+        thread::Builder::new()
```
1. I don't really understand this comment. What is "simple and direct thread creation", and how is that different than just "spawning a thread"? Also, why does it matter that the "current calls are infrequent"? And how do you know that they're infrequent, as that depends on how the application is using this crate?
2. Is it absolutely necessary to spawn an entire thread just to run the polkit check? If so, can you provide a link to some documentation (e.g., from the polkit docs) that recommend doing it on a separate thread?

I don't _mind_ the approach of doing this on a background thread, but it's a lot different than all other platforms, and typically we want to avoid spawning an entire thread just for one small request like this. However, it does make sense that an auth request would need to be done in the background, but it might also be unexpected since many other platforms require certain actions to be done on the main UI thread. I assume polkit does not require that it runs on the main thread (?).

tyreseluo replied:
Thanks for the review — I agree the comment was poorly worded and made assumptions that aren’t justified.

> On “simple and direct thread creation” and “infrequent calls”

You’re right: “simple and direct thread creation” is not meaningfully different from “spawning a thread”. What I meant was “we’re using std::thread::spawn rather than depending on an async runtime or a thread pool”, but the current wording is confusing. I’ll update the comment to be explicit and remove the “infrequent calls” justification — the crate cannot assume how frequently the application calls this API.

> Do we need a background thread for the polkit check?

It’s not that polkit requires a separate thread, but Authority.CheckAuthorization with AllowUserInteraction can block for an unbounded amount of time (D-Bus round-trips, waiting for a polkit agent, and potentially user interaction). If authenticate() is called on a UI/event-loop thread, doing the check synchronously would freeze that thread.

Because this crate exposes an async-style API via a callback on all platforms, the Linux implementation should avoid blocking the caller thread and keep similar semantics.

> Is there documentation that recommends a separate thread?

I don’t have a specific polkit doc quote that says “run this on a separate thread”. The rationale is general: CheckAuthorization may block and is performed over D-Bus; therefore it should not run on a UI thread. I’ll clarify that in the comment rather than implying it is a strict polkit requirement.

> Consistency with other platforms / main-thread requirement

Polkit does not require running on the main UI thread. In fact, running it on the main thread is undesirable if the call blocks. The callback is currently invoked on the background thread, and callers that need to update UI should dispatch back to their main/UI thread (similar to how many callback-based APIs behave).

I’ll update the implementation to run the polkit check off the caller thread and update the comment to clearly explain that this is done to avoid blocking the caller thread, not because we assume calls are infrequent. If thread-per-call overhead becomes a concern, we can follow up with a shared worker/thread-pool approach without changing the public API.

tyreseluo replied:
## Solution 1

```rust
pub(crate) fn authenticate<F>(
    &self,
    _message: Text,
    policy: &Policy,
    callback: F,
) -> Result<()>
where
    F: Fn(Result<()>) + Send + 'static,
{
    let action_id = policy.action_id;

    // CheckAuthorization may block (D-Bus round-trips and potential user interaction).
    // Run it off the caller thread to avoid blocking UI/event-loop threads.
    std::thread::Builder::new()
        .name("robius-authentication-polkit".into())
        .spawn(move || {
            let res = do_polkit_check(action_id);
            callback(res);
        })
        .map_err(|_| Error::Unavailable)?;

    Ok(())
}
```

## Solution 2

Single-worker thread queue also can solve this problem.

Sample code (simplified version):
```rust
use std::sync::{mpsc, OnceLock};

type Job = ( &'static str, Box<dyn FnOnce(Result<()>) + Send> );

static WORKER: OnceLock<mpsc::Sender<Job>> = OnceLock::new();

fn get_worker() -> mpsc::Sender<Job> {
    WORKER.get_or_init(|| {
        let (tx, rx) = mpsc::channel::<Job>();
        std::thread::Builder::new()
            .name("robius-authentication-polkit-worker".into())
            .spawn(move || {
                while let Ok((action_id, cb)) = rx.recv() {
                    let res = do_polkit_check(action_id);
                    cb(res);
                }
            })
            .expect("failed to spawn polkit worker");
        tx
    }).clone()
}

pub(crate) fn authenticate<F>(
    &self,
    _message: Text,
    policy: &Policy,
    callback: F,
) -> Result<()>
where
    F: Fn(Result<()>) + Send + 'static,
{
    let tx = get_worker();
    let action_id = policy.action_id;

    tx.send((action_id, Box::new(move |res| callback(res))))
        .map_err(|_| Error::Unavailable)?;

    Ok(())
}
```

kevinaboos replied:
Great, thanks for the detailed responses and investigation.

Spawning a new thread on demand (per call) is totally fine for this, authentication is human-time bound and is quite slow, so there a no need to worry about the latency. No need for a thread pool either.

kevinaboos commented on crates/authentication/src/sys/linux/mod.rs:
```diff
@@ -0,0 +1,169 @@
+}
+
+fn do_polkit_check(action_id: &'static str) -> Result<()> {
+    // Get authority proxy, and it's cached for future use.
```
```suggestion
    // Get authority proxy (and cache it for future usage).
```

kevinaboos commented on crates/authentication/src/sys/linux/mod.rs:
```diff
@@ -0,0 +1,169 @@
+    let subject = Subject::new_for_owner(pid, None, None)
+        .map_err(|_| Error::Unavailable)?;
+
+    // If details is non-empty then the request will fail with POLKIT_ERROR_FAILED unless the process doing the check itsef is sufficiently authorized (e.g. running as uid 0).
```
nit: split comment across multiple lines

kevinaboos commented on crates/authentication/src/sys/linux/mod.rs:
```diff
@@ -0,0 +1,169 @@
+    map_authorization_result(result)
+}
+
+fn map_authorization_result(result: AuthorizationResult) -> Result<()> {
```
no need to make this a separate function, just combine it into the above function.

kevinaboos commented on crates/authentication/src/sys/linux/mod.rs:
```diff
@@ -0,0 +1,169 @@
+    } else {
+        Error::Authentication
+    }
+}
```
surely this is not the best way of figuring out what the error is. What if the user's machine is in a different language other than english?
Typically you would use error codes or something instead of a string search.

kevinaboos commented on crates/authentication/src/sys/linux/mod.rs:
```diff
@@ -0,0 +1,169 @@
+        return Ok(());
+    }
+
+    if result.details.keys().any(|k| k.contains("dismissed")) {
```
surely this is not the best way of determining if the user dismissed the auth prompt. What happens if the user's machine is in a different language other than english?
Typically you would use error codes or something instead of a string search.

kevinaboos commented on crates/authentication/src/sys/linux/mod.rs:
```diff
@@ -0,0 +1,169 @@
+
+/// Policy builder for Linux.
+/// On Linux, a polkit action id MUST be explicitly provided.
+/// If not set, build() returns None.
```
like i said in a previous comment, this doc comment needs to go on the top-level `Policy` or `PolicyBuilder` type such that a user of this crate can see that information. This is a hidden inner module that the user will never ben able to view.

kevinaboos commented on crates/authentication/src/sys/linux/mod.rs:
```diff
@@ -0,0 +1,169 @@
+    pub const fn wrist_detection(self, _wrist: bool) -> Self { self }
+
+    /// Cross-platform API requirement: return Option.
+    /// Linux behavior: None if action_id is not explicitly set.
```
this comment also needs to be moved to the top-level `build()` function, or perhaps it's enough to just have the other doc comments on the `PolicyBuilder` type itself.

kevinaboos reviewed (changes requested):
Thanks Tyrese, I really appreciate this contribution!

I left a few comments about documentation style/location, but the most important questions I have are about the proper way to determine what kind of authentication error occurred. Doing a string search of the error message cannot possibly be the correct way.

tyreseluo committed:
Update crates/authentication/README.md

Co-authored-by: Kevin Boos <1139460+kevinaboos@users.noreply.github.com>

tyreseluo committed:
Update crates/authentication/README.md

Co-authored-by: Kevin Boos <1139460+kevinaboos@users.noreply.github.com>

tyreseluo committed:
Modify and fix some issues

tyreseluo committed:
Update crates/authentication/src/sys/linux/mod.rs

Co-authored-by: Kevin Boos <1139460+kevinaboos@users.noreply.github.com>

tyreseluo committed:
Add some comments

tyreseluo committed:
Merge remote-tracking branch 'origin/add_authentication_for_linux' into add_authentication_for_linux

tyreseluo committed:
Update and resolve some issue

tyreseluo committed:
Update polkit checks asynchronously and harden Linux auth flow

tyreseluo commented:
## Using authentication in Robrix on Linux
<img width="2576" height="1700" alt="image" src="https://github.com/user-attachments/assets/821e1252-f9fb-4b0a-b0e0-df41bad39dea" />

tyreseluo committed:
Improve Linux polkit docs, examples, and error handling

- Document Release mode policy installation with cargo-packager
- Clarify async auth flow in the CLI example
- Tighten polkit error mapping/logging and internal visibility

tyreseluo commented:
Hi! @kevinaboos I have resolved all issues and updated the Linux section of the README. You may review it again. Thank you very much.

kevinaboos commented:
Excellent, thanks! I will take a look shortly.

kevinaboos commented on crates/authentication/README.md:
```diff
@@ -30,43 +30,63 @@ To use this crate on Android, you must add the following to your app's `AndroidM
+
+### Write policy file.
+
+You can crate your own application's policy file, also can crate by template policy file.
```
```suggestion
You can create your application's own policy file from scratch, or also create one from a template policy file.
```

kevinaboos commented on crates/authentication/README.md:
```diff
@@ -30,43 +30,63 @@ To use this crate on Android, you must add the following to your app's `AndroidM
+
+You can crate your own application's policy file, also can crate by template policy file.
+
+Template policy file see: [`./examples/org.robius.authentication.policy`](./examples/org.robius.authentication.policy)
```
```suggestion
See here for an example template policy file: [`./examples/org.robius.authentication.policy`](./examples/org.robius.authentication.policy)
```

kevinaboos commented on crates/authentication/README.md:
```diff
@@ -30,43 +30,63 @@ To use this crate on Android, you must add the following to your app's `AndroidM
+
+### Quick Test Mode ⚠️
+
+Add the policy file to actions by manually executing the following command:
```
"add ... to actions"

what is "actions" here? i think we need more details to specify what "actions" refers to here.

kevinaboos commented on crates/authentication/README.md:
````diff
@@ -30,43 +30,63 @@ To use this crate on Android, you must add the following to your app's `AndroidM
+sudo install -Dm644 com.yourapp.policy /usr/share/polkit-1/actions/
+```
+
+Then, ensure your policy file is correctly installed.
````
nit
```suggestion
Then, ensure your policy file was correctly installed.
```

kevinaboos commented on crates/authentication/README.md:
```diff
@@ -30,43 +30,63 @@ To use this crate on Android, you must add the following to your app's `AndroidM
-which allows you to customize the prompt for each platform.
+> During the test mode, you don't need to worry about the location of the policy file; just ensure it installs correctly.
+>
+> You can also store it in advance under the unified packaging configuration folder (like `packaging`)to facilitate automatic installation of the policy file during release mode when users perform installation.
```
what does "store it in advance" mean? As in, create it in advance and move/install it into the proper directory _before_ installing the application? 

I feel like this sentence doesn't really help anyone; what were you hoping to convey here? Is this for developers or for users? or perhaps for people who are packaging up an app?

Missing a space after "(like `packaging`)"

kevinaboos commented on crates/authentication/src/lib.rs:
```diff
@@ -141,6 +141,8 @@ impl Context {
     /// Note that the callback may be not be called at all,
     /// but will always be called upon success.
+    /// 
+    /// On Linux: message is unused.
```
Add more explanation of _why_ it is unused on Linux, and what the alternative is.

(e.g., something like the auth messages are defined in your policy file, etc)

by the way, If an application wants to use _multiple_ different authentication message strings, how do you define that in a policy file? Would we need to allow the developer to select one of those defined message strings here when showing the auth prompt?

kevinaboos commented on crates/authentication/src/lib.rs:
```diff
@@ -178,6 +180,9 @@ pub enum BiometricStrength {
 /// will be displayed correctly on all platforms.
+///
+/// On Linux, a polkit action id MUST be explicitly provided.
+/// If not set, build() returns None. None if action_id is not explicitly set.
```
```suggestion
/// On Linux, a polkit action ID *MUST* be explicitly provided.
/// If not set, `build()` will return None.
```

kevinaboos commented on crates/authentication/src/lib.rs:
```diff
@@ -199,6 +204,24 @@ impl PolicyBuilder {
+    /// This only has an effect on linux.
+    #[inline]
+    #[must_use]
+    pub const fn action_id(self, id: &'static str) -> Self {
```
this only accepts a static str, which must be hardcoded as a constant value. Generally this is bad API practice because it prevents the developer from defining/determining a string dynamically at runtime.

Is there _any situation_ in which a developer might want to select or provide an `action_id` dynamically? 

I think you should probably change this to be a regular `String` just in case. (and apply that change to all places where the action ID is used/stored/passed around).

kevinaboos commented on crates/authentication/src/sys/linux/mod.rs:
```diff
@@ -0,0 +1,164 @@
+}
+
+static SYSTEM_CONNECTION: OnceLock<Result<Connection>> = OnceLock::new(); // Cached system bus connection
+static AUTHORITY_PROXY: OnceLock<Result<AuthorityProxyBlocking<'static>>> = OnceLock::new(); // Cached authority proxy
```
Question: is it safe/correct to cache these values, or do we need to re-obtain them every time we want to show a new auth prompt?

I only ask because many different system states (like channels/sockets) change frequently, so you shouldn't necessarily cache them because the old values might no longer be valid when you want to use them next time.

For ex, if the app runs for a very long time, perhaps across sleeping/hibernating of the OS, then things like dbus connections may not be valid. Please test this out.

In general, I'd probably recommend _not_ caching these values since it doesn't offer much benefit, and may behave differently/unexpectedly across varying Linux distros.

tyreseluo replied:
yes, i agree.

kevinaboos commented on crates/authentication/src/sys/linux/mod.rs:
```diff
@@ -0,0 +1,164 @@
+
+    if result
+        .details
+        .get("polkit.dismissed")
+        .map(|v| !v.is_empty())
+        .unwrap_or(false)
+    {
+        return Err(Error::UserCanceled);
+    }
+
+
+    // If we're not authorized, treat as authentication failure.
+    // "NoAgent" is handled as an error (org.freedesktop.PolicyKit1.Error.NoAgent).
+    Err(Error::Authentication)
+}
+
+#[derive(Debug)]
+pub struct Policy {
+    pub(crate) action_id: &'static str,
+}
+
+#[derive(Debug)]
+pub(crate) struct PolicyBuilder {
+    action_id: Option<&'static str>,
+}
```
as per my previous comment, it's better to just use a `String` here for future potential changes.

kevinaboos reviewed (changes requested):
I left a few comments, mostly questions/concerns about Linux system behavior and how it might violate some assumptions that you've made.

tyreseluo committed:
Update and fix some documents issues

tyreseluo committed:
Update authentication(linux) validate polkit action IDs

- replace action_id with action_ids allowlist
- add InvalidActionId and Linux validation
- update examples and policy template

tyreseluo committed:
Authentication/linux: avoid caching dbus connection and polkit proxy

tyreseluo committed:
Authentication: explain why Linux ignores message text

tyreseluo commented:
All problems was resolved, I've grown. Thanks @kevinaboos! You can review again 🤝🤝🤝

kevinaboos committed:
Merge branch 'main' into add_authentication_for_linux

kevinaboos reviewed (approved):
Thanks again Tyrese, apologies it took so long for me to get on top of this.
